apollo-integration/src/test/java/com/apollographql/apollo/ApolloPrefetchTest.java

```diff
@@ -73,137 +70,65 @@ public void setup() {
-    assertThat(apolloClient.query(new AllPlanetsQuery())
-        .httpCachePolicy(HttpCachePolicy.CACHE_ONLY.expireAfter(2, TimeUnit.SECONDS)).execute()
-        .hasErrors()).isFalse();
+    assertResponse(
+        apolloClient
+            .query(new AllPlanetsQuery())
+            .httpCachePolicy(HttpCachePolicy.CACHE_ONLY.expireAfter(2, TimeUnit.SECONDS)),
+        new Predicate<Response<AllPlanetsQuery.Data>>() {
+          @Override public boolean test(Response<AllPlanetsQuery.Data> dataResponse) throws Exception {
+            return !dataResponse.hasErrors();
+          }
+        }
+    );
   }
 
-  @Test public void prefetchNoCacheStore() throws IOException, ApolloException {
+  @Test public void prefetchNoCacheStore() throws Exception {
     ApolloClient apolloClient = ApolloClient.builder()
         .serverUrl(server.url("/"))
         .okHttpClient(okHttpClient)
         .build();
 
-    enqueueResponse("HttpCacheTestAllPlanets.json");
-    apolloClient.prefetch(new AllPlanetsQuery()).execute();
-    enqueueResponse("HttpCacheTestAllPlanets.json");
-    assertThat(apolloClient.query(new AllPlanetsQuery()).execute().hasErrors()).isFalse();
+    server.enqueue(mockResponse("HttpCacheTestAllPlanets.json"));
+    awaitDone(apolloClient.prefetch(new AllPlanetsQuery()));
+
+    enqueueAndAssertResponse(
+        server,
+        "HttpCacheTestAllPlanets.json",
+        apolloClient.query(new AllPlanetsQuery()),
+        new Predicate<Response<AllPlanetsQuery.Data>>() {
+          @Override public boolean test(Response<AllPlanetsQuery.Data> response) throws Exception {
+            return !response.hasErrors();
+          }
+        }
+    );
   }
 
   @Test public void prefetchFileSystemWriteFailure() throws IOException {
     FaultyHttpCacheStore faultyCacheStore = new FaultyHttpCacheStore(FileSystem.SYSTEM);
     cacheStore.delegate = faultyCacheStore;
 
-    enqueueResponse("HttpCacheTestAllPlanets.json");
     faultyCacheStore.failStrategy(FaultyHttpCacheStore.FailStrategy.FAIL_HEADER_WRITE);
-    try {
-      apolloClient.prefetch(new AllPlanetsQuery()).execute();
-      fail("exception expected");
-    } catch (Exception expected) {
-    }
+    server.enqueue(mockResponse("HttpCacheTestAllPlanets.json"));
+
+    Rx2Apollo.from(apolloClient.prefetch(new AllPlanetsQuery()))
+        .test()
+        .awaitDone(TIME_OUT_SECONDS, TimeUnit.SECONDS)
+        .assertError(Exception.class);
     checkNoCachedResponse();
 
-    enqueueResponse("HttpCacheTestAllPlanets.json");
+    server.enqueue(mockResponse("HttpCacheTestAllPlanets.json"));
     faultyCacheStore.failStrategy(FaultyHttpCacheStore.FailStrategy.FAIL_BODY_WRITE);
-    try {
-      apolloClient.prefetch(new AllPlanetsQuery()).execute();
-      fail("exception expected");
-    } catch (Exception expected) {
-    }
+    Rx2Apollo.from(apolloClient.prefetch(new AllPlanetsQuery()))
+        .test()
+        .awaitDone(TIME_OUT_SECONDS, TimeUnit.SECONDS)
```

Comment:
I usually use `awaitTerminalEvent` in cases like this, not sure what difference is with `awaitDone`. From what I gather `awaitTerminalEvent` waits for `onComplete` or `onError` to be called while `awaitDone` merely waits for latch.

Reply:
there is not much difference:
```
public final U awaitDone(long time, TimeUnit unit) {
        try {
            if (!done.await(time, unit)) {
                dispose();
            }
        } catch (InterruptedException ex) {
            dispose();
            throw ExceptionHelper.wrapOrThrow(ex);
        }
        return (U)this;
    }
```
and 
```
public final boolean await(long time, TimeUnit unit) throws InterruptedException {
        return done.getCount() == 0 || done.await(time, unit);
    }

public final boolean awaitTerminalEvent(long duration, TimeUnit unit) {
        try {
            return await(duration, unit);
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            return false;
        }
    }
```

So the first API is just call dispose on timeout and it returns the the TestObserver so you can chain asserts where second just awaits and returns boolean